File: scanners/headless/local_bridge.js
#!/usr/bin/env node

/******
* Part of a Python<->Node bridge for local scans.
*
* Executed by local_bridge.py, to send scan parameters and
* receive scan responses over the CLI and STDOUT.
*******/

const puppeteer = require('puppeteer');
var base = require("./base")

// the first argument passed in is the name of the scanner
const scanner = require("../" + process.argv[2]);

// the second argument passed is JSON-serialized data
const params = JSON.parse(process.argv[3]);

// Async function to load Chrome from the local system.
var getBrowser = async () => {
  return await puppeteer.launch({
    // TODO: Let executable path be overrideable.
    // executablePath: config.executablePath,
    headless: true,
    // ignoreHTTPSErrors: true,
    args: [
      '--no-sandbox',
      '--disable-gpu',
      '--single-process'
    ]
  });
};

var go = async () => {

  // Execute the scan using the passed-in data, and the
  // locally launched headless Chrome browser.
  base.scan(
    params.domain, params.environment, params.options,
    getBrowser, scanner,
    function(err, data) {
      if (err) {
        console.error(err);
        process.exit(1);
      }
      console.log(JSON.stringify(data))
    }
  );

};

go();

File: scanners/headless/local_bridge.py
import logging
import json

from utils import scan_utils

###
# Local Python bridge to the JS bridge to the JS scanner.
# Shells out to run Node on base.js, which runs headless Chrome
# and then pulls in the scanner-specific JS scanning function.
# Resulting data is serialized and returned via STDOUT.
###


def headless_scan(scanner_name, domain, environment, options):
    raw = scan_utils.scan(
        [
            "./scanners/headless/local_bridge.js",
            scanner_name,
            scan_utils.json_for({
                'domain': domain,
                'environment': environment,
                'options': options
            })
        ]
    )

    if not raw:
        logging.warning("\tError calling out to %s.js, skipping." % scanner_name)
        return None

    try:
        data = scan_utils.from_json(raw)
    except json.decoder.JSONDecodeError:
        logging.warning("\tError inside %s.js, skipping. Error below:\n\n%s" % (scanner_name, raw))
        return None

    return data
